File: php/src/whatsapp.php
<?php

declare(strict_types=1);

/** @return array<string,mixed> */
function whatsapp_parse_event(array $payload): array
{
    $value = $payload['entry'][0]['changes'][0]['value'] ?? $payload; $message = $value['messages'][0] ?? $payload['message'] ?? []; $user = $message['from'] ?? ''; if ($user === '') throw new InvalidArgumentException('WhatsApp webhook has no sender'); $raw = $message['type'] ?? 'text'; $type = $raw === 'document' ? 'file' : $raw; $item = $message[$raw === 'document' ? 'document' : $raw] ?? []; return ['platform'=>'whatsapp','user_id'=>(string)$user,'content_type'=>$type,'text'=>$item['body'] ?? $item['caption'] ?? null,'media_url'=>$item['id'] ?? null];
}
function whatsapp_send_reply(array $event, array $reply, string $phoneNumberId, string $token): void
{
    if ($phoneNumberId === '' || $token === '') throw new InvalidArgumentException('WhatsApp phone number ID and token are required'); foreach (array_slice($reply['messages'] ?? [],0,5) as $message) { $type = ($message['type'] ?? 'text') === 'file' ? 'document' : ($message['type'] ?? 'text'); $body = ['messaging_product'=>'whatsapp','to'=>$event['user_id'],'type'=>$type]; if ($type === 'text') $body['text']=['body'=>$message['text'] ?? '']; else { if (empty($message['media_url'])) throw new InvalidArgumentException('WhatsApp media reply requires media ID'); $body[$type]=['id'=>$message['media_url']]; } $ch=adapter_curl_init('https://graph.facebook.com/v20.0/'.$phoneNumberId.'/messages');curl_setopt_array($ch,[CURLOPT_POST=>true,CURLOPT_RETURNTRANSFER=>true,CURLOPT_HTTPHEADER=>['Authorization: Bearer '.$token,'Content-Type: application/json'],CURLOPT_POSTFIELDS=>json_encode($body,JSON_THROW_ON_ERROR)]);$result=curl_exec($ch);$status=curl_getinfo($ch,CURLINFO_RESPONSE_CODE);if($result===false||$status<200||$status>=300)throw new RuntimeException('WhatsApp API request failed');curl_close($ch); }
}

File: python/platforms/whatsapp.py
from __future__ import annotations
import os
from typing import Any
import requests
from core.models import InboundEvent, OutboundReply
from .base import PlatformAdapter
from .catalog import PLATFORM_CATALOG

class WhatsAppAdapter(PlatformAdapter):
    def __init__(self, token: str | None = None, phone_number_id: str | None = None, session: requests.Session | None = None):
        self.capabilities = PLATFORM_CATALOG["whatsapp"]; self.token = token or os.getenv("WHATSAPP_ACCESS_TOKEN"); self.phone_number_id = phone_number_id or os.getenv("WHATSAPP_PHONE_NUMBER_ID")
        if not self.token or not self.phone_number_id: raise ValueError("WHATSAPP_ACCESS_TOKEN and WHATSAPP_PHONE_NUMBER_ID are required")
        self.session = session or requests.Session()
    def parse_event(self, payload: Any, headers=None) -> InboundEvent:
        value = (payload.get("entry") or [{}])[0].get("changes", [{}])[0].get("value", payload); message = (value.get("messages") or [{}])[0]
        kind = message.get("type", "text")
        kind = {"text": "text", "image": "image", "audio": "audio", "video": "video", "document": "file"}.get(kind, "text")
        item = message.get("document" if kind == "file" else kind, {})
        return InboundEvent(platform="whatsapp", user_id=str(message.get("from") or payload.get("user_id")), content_type=kind, text=item.get("body") or item.get("caption"), media_url=item.get("id"))
    def send_reply(self, event: InboundEvent, reply: OutboundReply) -> None:
        url = f"https://graph.facebook.com/v20.0/{self.phone_number_id}/messages"
        for message in reply.messages:
            body = {"messaging_product": "whatsapp", "to": event.user_id}
            if message.type == "text":
                body.update({"type": "text", "text": {"body": message.text}})
            else:
                media_key = {"image": "image", "audio": "audio", "video": "video", "file": "document"}[message.type]
                if not message.media_url:
                    raise ValueError(f"WhatsApp {message.type} reply requires media_url")
                body["type"] = media_key
                body[media_key] = {"id": message.media_url, **({"caption": message.text} if message.text and media_key in ("image", "video", "document") else {})}
            response = self.session.post(url, headers={"Authorization": f"Bearer {self.token}"}, json=body, timeout=30); response.raise_for_status()
